### PR TITLE
[AUD-1108] CI Improvements

### PR DESCRIPTION
**Pull request by sliptype**
### Description

This PR:
* Limits the ci build-upload workflow to release branches
* Adds auto incrementing of version name and build number. See notes below
* Builds the client in the ci workflow instead of pulling in an existing bundle from s3
* Adds lint and typecheck to the init step

Things I still need to do (after soaking this change for a bit):
* Update docs for releasing mobile
* Remove bump build number command from package.json
* Remove mobile bundle build from client ci workflow

Version increment notes:

  Version names and build numbers behave differently for android and ios:

  Android - Build number is unique, always incrementing. Version name is just a string to show the user.
  For android, ci increments both the version name and build number based on what already exists in google play

  iOS - Version name must be unique, build number must be unique for that version name. 
  For ios, ci increments the version name and then increments the existing build number for that version name. It defaults to 1 
  if the version doesn't exist yet

  For both android and ios, the version name can be updated in the source code. If the version name is higher than what 
  already exists, the specified version name will be used

### Dragons

*Is there anything the reviewer should be on the lookout for? Are there any dangerous changes?*

### How Has This Been Tested?

*Please describe the tests that you ran to verify your changes. Provide repro instructions & any configuration.*

### How will this change be monitored?

*For features that are critical or could fail silently please describe the monitoring/alerting being added.*

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -146,6 +180,9 @@ jobs:
       - checkout
       - attach_workspace:
           at: ./
+      - run:
+          name: Move build
```
Comment (raymondjacobson) [inline .circleci/config.yml:184]: `Move Production build`

just so folks see the steps and know *for sure* what's happening

File: .circleci/config.yml [hunk 2]
```diff
@@ -224,38 +267,47 @@ jobs:
...
-          name: migrate support libraried for androidX packages
-          command: |
-            npm run jetifier
+          name: Set Ruby Version
+          command: echo 'chruby ruby-2.6' >> ~/.bash_profile
       - run:
-          name: generating the release apk & bundle
-          command: |
-            echo "yes" | $ANDROID_HOME/tools/bin/sdkmanager "platforms;android-30" && cd android && ./gradlew bundleRelease
+          name: Move build
```
Comment (raymondjacobson) [inline .circleci/config.yml:279]: same here

File: android/fastlane/Fastfile [hunk 3]
```diff
@@ -61,4 +61,51 @@ platform :android do
...
+
+  lane :incrementVersionCode do |options|
+      # Get the max existing version code
+      previousVersion = google_play_track_version_codes(
+        package_name: options[:package_name],
+        track: options[:track]
+      ).max
+
+      # Increment the version code in build.gradle by 1
+      re = /versionCode\s+(\d+)/
```
Comment (raymondjacobson) [inline android/fastlane/Fastfile:74]: ouch

**Review by raymondjacobson — APPROVED** [review]
This is all of the things!!!!

**Review by rickyrombo — APPROVED** [review]
Love this! Not sure I fully understand the nuance with the buildnumbers - in both android and iOS the version name and build number are incremented right?

**Comment by sliptype** [comment]
> Love this! Not sure I fully understand the nuance with the buildnumbers - in both android and iOS the version name and build number are incremented right?

It's confusing but essentially yes, both are incremented. I'm working on a doc to explain it: https://www.notion.so/audiusproject/Mobile-Client-Versioning-e702b0a36c924c7abe2a40a90d99169d

**Pull request merged**